### PR TITLE
feat: add File.force_close to close bytesio files

### PR DESCRIPTION
**Pull request by TrapDrap**
## Summary

Added force_close to `File` class that will prevent memory leak to occur when a message containing a file is sent if enabled.
`nextcord.abc` calls `File.close()` for each file after the message is sent.
Files bytes could not be deleted after `File.close()` is called.
Causing uploaded files to linger in memory until the garbage collector removes them, usually after a few hours.

## Checklist

- [x] If code changes were made then they have been tested.
  - [x] I have updated the documentation to reflect the changes.
  - [x] I have run `task pyright` and fixed the relevant issues.
- [x] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Review by Skelmis — APPROVED** [review]
Untested, however looks good to me

**Review by EmreTech — APPROVED** [review]

**Comment by DenverCoder1** [comment]
I feel like the `_owner` check must have been there for some reason.

Are you sure this is not reverting work done in https://github.com/yagomichalak/discord.py/commit/f55cd18b494a6044666357e4f1daa3ebcb5e4afe to fix https://github.com/Rapptz/discord.py/issues/1809?

Maybe a 2-step close on your end would avoid potentially causing issues with it being closed when it is not supposed to?
```py
file.close()
file.fp.close()
```

(I haven't tested it yet, so I'm still not sure if the check is required or not.)

**Comment by TrapDrap** [comment]
> 

It shouldn't.
aiohttp shouldn't call the `.close()` method from of the File class, it calls `.fp.close()` from the BufferIO/BytesIO object, which was stubbed in line 114 & 115.
I did not remove the stub that revoked the ability for aiohttp to close the data.

**Comment by DenverCoder1** [comment]
After calling file.close, the fp.close method is restored to its original method on the first line

**Comment by TrapDrap** [comment]
> After calling file.close, the fp.close method is restored to its original method on the first line

But question is why make the user have to manually close the file, it should be handled on file closing
You also wouldn't be able to do that if it was sent something like `await ch.send(file=File(bytes, name))`

**Comment by DenverCoder1** [comment]
Here's what discord.py says about it;

> if the File object doesn't own the io object (it owns it e.g. when you supply a filename), then it assumes you'll be the one to handle closing it

\- Zephyrkul 

> yup
> (also, closing a BytesIO doesn't really do anything tangible)

\- Danny

**Comment by TrapDrap** [comment]
> > if the File object doesn't own the io object (it owns it e.g. when you supply a filename), then it assumes you'll be the one to handle closing it

This also applies to converting an asset to file btw. (Using `.to_file()`)

> > Here's what discord.py says about it

What is the point of keeping it the same? We don't have to be the same as dpy 

What if we gave `File` a parameter for automatic closing, defaulting to True?
In most cases, you won't need to reuse bytes that have been sent.

**Comment by DenverCoder1** [comment]
I'm not saying it has to be the same as discord.py, I'm just adding it here for additional context so it can be discussed.

I do think there is a case for keeping it as is, for example if you are using BytesIO for a running buffer and want to keep appending to it even after a snapshot has been sent as a file.

I like your idea of making it togglable, though, especially for the case you mentioned with to_file where the user may be unaware the BytesIO is not closed automatically.

Maybe to avoid breaking changes, it could default to the current behavior, but override the _owner check if the new attribute (force_close?) is enabled?

**Comment by TrapDrap** [comment]
> Maybe to avoid breaking changes, it could default to the current behavior, but override the _owner check if the new attribute (force_close?) is enabled?

That sounds like a good idea, I will work on it tmr

File: nextcord/file.py [hunk 1]
```diff
@@ -144,6 +167,7 @@ def reset(self, *, seek: Union[int, bool] = True) -> None:
             self.fp.seek(self._original_pos)
 
     def close(self) -> None:
-        self.fp.close = self._closer
-        if self._owner:
+        if self._owner or self.force_close:
             self._closer()
+        else:
+            self.fp.close = self._closer
```
Comment (DenverCoder1) [inline nextcord/file.py:173]: `self.fp.close = self._closer` should probably stay where it is. Keeping the order the same and just adding `or self.force_close` would be the best to avoid unnecessary changes.

File: nextcord/file.py [hunk 2]
```diff
@@ -95,6 +117,7 @@ def __init__(
         *,
         description: Optional[str] = None,
         spoiler: bool = False,
+        force_close: bool = True,  # Accessed in .close()
```
[the remark this answers is not in the document]
Comment (TrapDrap) [reply nextcord/file.py:120]: I believe it should default to `True` since there are rarely any cases where you would need persistent bytes after using the file.
But I'm open to getting more opinions from others too.

File: nextcord/file.py (hunk 1 above)
Comment (TrapDrap) [reply nextcord/file.py]: done

File: nextcord/file.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ooliver1) [reply nextcord/file.py:120]: As the discussion you guys have had above says, keeping it to the original behaviour avoids breaking changes (user closing it manually may throw an error)

[the remark this answers is not in the document]
Comment (TrapDrap) [reply nextcord/file.py:120]: > As the discussion you guys have had above says, keeping it to the original behaviour avoids breaking changes (user closing it manually may throw an error)

The concern isn't that. The user should be able to call `.close()` multiple times without a problem. That includes if the file is automatically closed. The concern lies only if the user plans to reuse the file bytes for whatever reason.
But with due respect, since you both think it's a better idea to keep `force_close` False for user-created files.

**Review by DenverCoder1 — APPROVED** [review]

File: nextcord/file.py [hunk 3]
```diff
@@ -63,6 +63,12 @@ class File:
         in the Discord client.
     spoiler: :class:`bool`
         Whether the attachment is a spoiler.
+    force_close: :class:`bool`
+        Whether to forcibly close the bytes used to create the file
+        when ``.close()`` is called.
+        This will also make the file bytes unusable by flusing it from
```
Comment (DenverCoder1) [inline nextcord/file.py:69]:
```suggestion
        This will also make the file bytes unusable by flushing it from
```

File: nextcord/file.py [hunk 4]
```diff
@@ -63,6 +63,12 @@ class File:
...
+        memory after it is sent once.
+        Keep this enabled if you don't wish to reuse the same bytes.
```
Comment (DenverCoder1) [inline nextcord/file.py:71]:
```suggestion
        Enable this if you don't wish to reuse the same bytes.
```

File: nextcord/file.py [hunk 5]
```diff
@@ -78,15 +84,31 @@ class File:
         in the Discord client.
     spoiler: :class:`bool`
         Whether the attachment is a spoiler.
+    force_close: :class:`bool`
+        Whether to forcibly close the bytes used to create the file
+        when ``.close()`` is called.
+        This will also make the file bytes unusable by flusing it from
+        memory after it is sent or used once.
+        Keep this enabled if you don't wish to reuse the same bytes.
```
Comment (DenverCoder1) [inline nextcord/file.py:92]:
```suggestion
        Whether to forcibly close the bytes used to create the file
        when ``.close()`` is called.
        This will also make the file bytes unusable by flushing it from
        memory after it is sent or used once.
        Enable this if you don't wish to reuse the same bytes.
```

**Review by DenverCoder1 — CHANGES_REQUESTED** [review]

**Comment by DenverCoder1** [comment]
If you want to set it to True in the `to_file` functions, I don't think there would be any harm as far as I can tell

**Comment by TrapDrap** [comment]
> If you want to set it to True in the `to_file` functions, I don't think there would be any harm as far as I can tell

Yes I am working on that currently

File: nextcord/asset.py [hunk 6]
```diff
@@ -141,6 +142,12 @@ async def to_file(
             The description for the file.
         spoiler: :class:`bool`
             Whether the file is a spoiler.
+        force_close: :class:`bool`
+            Whether to forcibly close the bytes used to create the file
+            when ``.close()`` is called.
+            This will also make the file bytes unusable by flusing it from
```
Comment (DenverCoder1) [inline nextcord/asset.py:148]:
```suggestion
            This will also make the file bytes unusable by flushing it from
```

File: nextcord/file.py (hunk 3 above)
Comment (DenverCoder1) [reply nextcord/file.py:69]: Was marked as resolved, but it doesn't seem like these were fixed yet

Comment (TrapDrap) [reply nextcord/file.py:69]: Weird I committed the latest

**Review by DenverCoder1 — APPROVED** [review]

**Review by Skelmis — APPROVED** [review]

File: nextcord/asset.py [hunk 7]
```diff
@@ -141,6 +142,12 @@ async def to_file(
             The description for the file.
         spoiler: :class:`bool`
             Whether the file is a spoiler.
+        force_close: :class:`bool`
+            Whether to forcibly close the bytes used to create the file
+            when ``.close()`` is called.
+            This will also make the file bytes unusable by flushing it from
+            memory after it is sent or used once.
+            Keep this enabled if you don't wish to reuse the same bytes.
```
Comment (ooliver1) [inline nextcord/asset.py:150]:
```suggestion
            Keep this enabled if you don't wish to reuse the same bytes.

            .. versionadded:: 2.2
```

File: nextcord/file.py [hunk 8]
```diff
@@ -63,6 +63,12 @@ class File:
         in the Discord client.
     spoiler: :class:`bool`
         Whether the attachment is a spoiler.
+    force_close: :class:`bool`
+        Whether to forcibly close the bytes used to create the file
+        when ``.close()`` is called.
+        This will also make the file bytes unusable by flushing it from
+        memory after it is sent once.
+        Enable this if you don't wish to reuse the same bytes.
```
Comment (ooliver1) [inline nextcord/file.py:71]:
```suggestion
        Enable this if you don't wish to reuse the same bytes.

        .. versionadded:: 2.2
```

File: nextcord/file.py [hunk 9]
```diff
@@ -78,15 +84,31 @@ class File:
         in the Discord client.
     spoiler: :class:`bool`
         Whether the attachment is a spoiler.
+    force_close: :class:`bool`
+        Whether to forcibly close the bytes used to create the file
+        when ``.close()`` is called.
+        This will also make the file bytes unusable by flushing it from
+        memory after it is sent or used once.
+        Enable this if you don't wish to reuse the same bytes.
```
Comment (ooliver1) [inline nextcord/file.py:92]:
```suggestion
        Enable this if you don't wish to reuse the same bytes.

        .. versionadded:: 2.2
```

**Review by ooliver1 — CHANGES_REQUESTED** [review]

**Review by ooliver1 — APPROVED** [review]

**Review by teaishealthy — APPROVED** [review]
